### PR TITLE
Problem Suggestion: Add "Spirale" to Bronze - Simulation

### PR DESCRIPTION
**Pull request by jeffkmeng**
User `j2X6cR7FHBfbkCschVmILyVzQA23` suggested adding the problem [Spirale](https://dmoj.ca/problem/coci17c5p2) to the `easier` table of the module [Bronze - Simulation](https://usaco.guide/bronze/simulation#problemlist-easier).

**Automatically Generated JSON:**
```json
{
  "uniqueId": "dmoj-spirale",
  "name": "Spirale",
  "url": "https://dmoj.ca/problem/coci17c5p2",
  "source": "DMOJ",
  "difficulty": "Normal",
  "isStarred": false,
  "tags": [
    "Complete Search"
  ],
  "solutionMetadata": {
    "kind": "autogen-label-from-site",
    "site": "DMOJ"
  }
}
```
**Additional Notes**:
This problem might be hard for bronze-level, but I found it about medium level.

*This PR was automatically generated from a user-submitted problem suggestion on the USACO guide.*

### CODE REVIEW

File: content/2_Bronze/Simulation.problems.json [hunk 1]
```diff
@@ -93,15 +105,16 @@
...
-      "uniqueId": "usaco-856",
-      "name": "The Bucket List",
-      "url": "http://www.usaco.org/index.php?page=viewproblem2&cpid=856",
-      "source": "Bronze",
-      "difficulty": "Easy",
+      "uniqueId": "dmoj-spirale",
+      "name": "Spirale",
+      "url": "https://dmoj.ca/problem/coci17c5p2",
+      "source": "DMOJ",
+      "difficulty": "Normal",
```
Comment (Farrius) [inline content/2_Bronze/Simulation.problems.json:112]: I would say for bronze is at least hard.

**Comment by Farrius** [comment]
Btw, can you add the solution since I didn't find it on the internet. Here is mine, but add yours if you want
`#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using ld = long double;

vector<int> ui = {0, -1, 0, 1}, uj = {1, 0, -1, 0}, di = {0, -1, 0, 1}, dj = {-1, 0, 1, 0};
int mat[50][50], cnt = 0, n, m, k, cur = 0;

void fn (int x, int y) {
	if (x < 0 || x >= m || y < 0 || y >= n) return;
	cnt++;
	mat[y][x] = min(mat[y][x], cur);
}

int main () {
	cin >> n >> m >> k;
	int x, y, z;
	vector<int> ti, tj;
	for (int i = 0; i < 50; i++) {
		for (int j = 0; j < 50; j++) {
			mat[i][j] = 1e9;
		}
	}
	for (int i = 0; i < k; i++) {
		cin >> x >> y >> z;
		x--; y--;
		swap(x, y);
		if (z == 0) {
			ti = di;
			tj = dj;
		} else {
			ti = ui;
			tj = uj;
		}
		int kk = 1;
		int di = 1;
		int c = 2;
		cur = 1;
		cnt = 0;
		fn(x, y);
		while (true) {
			if (cnt >= n * m) break;
			if (c == 0) {
				di++;
				c = 2;
			}
			for (int va = 0; va < di; va++) {
				x += tj[kk];
				y += ti[kk];
				cur++;
				fn(x, y);
			}
			c--;
			kk = (kk + 1) % 4;
		}
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cout << mat[i][j] << (j == m - 1 ? '\n' : ' ');
		}
	}
}


`

**Comment by Farrius** [comment]
Btw, can you add the solution since I didn't find it on the internet. Here is mine, but add yours if you want
```
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using ld = long double;

vector ui = {0, -1, 0, 1}, uj = {1, 0, -1, 0}, di = {0, -1, 0, 1}, dj = {-1, 0, 1, 0};
int mat[50][50], cnt = 0, n, m, k, cur = 0;

void fn (int x, int y) {
if (x < 0 || x >= m || y < 0 || y >= n) return;
cnt++;
mat[y][x] = min(mat[y][x], cur);
}

int main () {
cin >> n >> m >> k;
int x, y, z;
vector ti, tj;
for (int i = 0; i < 50; i++) {
for (int j = 0; j < 50; j++) {
mat[i][j] = 1e9;
}
}
for (int i = 0; i < k; i++) {
cin >> x >> y >> z;
x--; y--;
swap(x, y);
if (z == 0) {
ti = di;
tj = dj;
} else {
ti = ui;
tj = uj;
}
int kk = 1;
int di = 1;
int c = 2;
cur = 1;
cnt = 0;
fn(x, y);
while (true) {
if (cnt >= n * m) break;
if (c == 0) {
di++;
c = 2;
}
for (int va = 0; va < di; va++) {
x += tj[kk];
y += ti[kk];
cur++;
fn(x, y);
}
c--;
kk = (kk + 1) % 4;
}
}
for (int i = 0; i < n; i++) {
for (int j = 0; j < m; j++) {
cout << mat[i][j] << (j == m - 1 ? '\n' : ' ');
}
```
}
}

**Comment by Farrius** [comment]
Btw, can you add the solution since I didn't find it on the internet. Here is mine, but add yours if you want
```cpp
#include <bits/stdc++.h>
int main () {
}
```

**Comment by maggie-j-liu** [comment]
@Farrius 
> *This PR was automatically generated from a user-submitted problem suggestion on the USACO guide.*

so the person who submitted this suggestion is probably not watching this pr. you can make the changes you want, then ask someone to merge...

**Comment by Farrius** [comment]
@maggie-j-liu Thanks for your answer :) I didn't know that. Do you have any suggestions about the difficulty?

**Comment by maggie-j-liu** [comment]
I haven't tried this problem, but I think hard is fine?

**Comment by Genius3435** [comment]
My code (cleaned up):

```cpp
#include <bits/stdc++.h>
using namespace std;

const int di[4] = {0, 1, 0, -1};
const int dj[4] = {1, 0, -1, 0};
// right, down, left, up


const int maxn = 52;
int a[maxn][maxn];


const int maxlen = 8 * maxn;
// If we spend over 8 * max(n,m) steps
// outside of the grid, then we can stop


int main() {
	int N, M, K;
	cin >> N >> M >> K;

	for (int i = 0; i < N; ++i)
		for (int j = 0; j < M; ++j)
			a[i][j] = 2e9;
	
	for (int k = 0; k < K; ++k) {
		int i, j, t;
		cin >> i >> j >> t;
		t = t ? 3 : 1;
		--i; --j;
		
		int dir = 3, dis = 1, d = 1;
		for (int cnt = 0; cnt < maxlen; ) {
			// Move dis steps forward
			for (int l = 0; l < dis; ++l) {
				// If it's in the grid, add that to the min distances
				if (0 <= i && i < N && 0 <= j && j < M) {
					a[i][j] = min(a[i][j], d);
					cnt = 0;
				}
				
				i += di[dir];
				j += dj[dir];
				++d; ++cnt;
				
				if (cnt > maxlen) break;
			}

			// Turn CW or CCW
			dir = (dir+t)%4;

			// Add one to the distance every two 
			// turns (a cool trick to make a spiral)
			if (dir&1) dis++;
		}
	}

	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < M; ++j)
			cout << a[i][j] << ' ';
		cout << '\n';
	}
}
```

Should I add it somewhere else too?

**Comment by Genius3435** [comment]
And you can probably get around the `di`, `dj` thing by using if statements instead of `i += di[dir]; j += dj[dir];` if you wanted, although I'm not sure that's a good idea.

**Comment by maggie-j-liu** [comment]
> Should I add it somewhere else too?

Since this problem suggestion has been merged, you can add your solution if it's different/better than Farrius's (https://github.com/cpinitiative/usaco-guide/pull/1526#issuecomment-841642638). The [adding solutions](https://usaco.guide/general/adding-solution) module gives instructions for adding a solution. If your solution is similar, you don't need to add it.

**Pull request merged**